kmod/src/omap.c

```diff
@@ -0,0 +1,1062 @@
+			if (free_rid(&req->rids, entry) == 0)
+				send_response = true;
+		}
+		spin_unlock(&req->lock);
+
+		if (send_response) {
+			resp_rid = req->client_rid;
+			resp_id = req->client_id;
+			memcpy(map, &req->map, sizeof(struct scoutfs_open_ino_map));
+			rhashtable_remove_fast(&ominf->req_ht, &req->ht_head, req_ht_params);
+			call_rcu(&req->rcu, free_req_rcu);
+		}
+	}
+	rcu_read_unlock();
+
+	if (send_response)
+		ret = scoutfs_server_send_omap_response(sb, resp_rid, resp_id, map, 0);
+	else
+		ret = 0;
+	kfree(map);
+out:
+	return ret;
+}
+
+/*
+ * The server is shutting down.  Free all the server state associated
+ * with ongoing request processing.  Clients who still have requests
+ * pending will resend them to the next server.
+ */
+void scoutfs_omap_server_shutdown(struct super_block *sb)
+{
+	DECLARE_OMAP_INFO(sb, ominf);
+	struct rhashtable_iter iter;
+	struct llist_node *requests;
+	struct omap_request *req;
+	struct omap_request *tmp;
+
+	rhashtable_walk_enter(&ominf->req_ht, &iter);
+	rhashtable_walk_start(&iter);
+
+	for (;;) {
+		req = rhashtable_walk_next(&iter);
+		if (req == NULL)
+			break;
+		if (req == ERR_PTR(-EAGAIN))
+			continue;
+
+		if (req->rids.nr_rids != 0) {
+			free_rids(&req->rids);
+			rhashtable_remove_fast(&ominf->req_ht, &req->ht_head, req_ht_params);
+			call_rcu(&req->rcu, free_req_rcu);
+		}
+	}
+
+	rhashtable_walk_stop(&iter);
+	rhashtable_walk_exit(&iter);
+
+	requests = llist_del_all(&ominf->requests);
+	llist_for_each_entry_safe(req, tmp, requests, llnode)
+		kfree(req);
+
+	synchronize_rcu();
+}
+
+/*
+ * Make sure that we have an omap_lock for the cluster lock.  Multiple
+ * users can race to allocate a new olk and once its stored our use of
+ * the cluster lock stops it from being freed.
+ */
+static int get_omap_lock(struct scoutfs_lock *lock, struct scoutfs_omap_lock **olk_ret)
+{
+	struct scoutfs_omap_lock *olk = NULL;
+	int ret = 0;
+
+	spin_lock(&lock->omap_spinlock);
+	olk = lock->omap_lock;
+	spin_unlock(&lock->omap_spinlock);
+
+	if (olk == NULL) {
+		olk = kzalloc(sizeof(struct scoutfs_omap_lock), GFP_NOFS);
+		if (olk) {
+			olk->version = lock->write_version - 1; /* ensure refresh */
+			init_waitqueue_head(&olk->waitq);
+
+			spin_lock(&lock->omap_spinlock);
+			if (lock->omap_lock == NULL) {
+				lock->omap_lock = olk;
+			} else {
+				kfree(olk);
+				olk = lock->omap_lock;
+			}
+			spin_unlock(&lock->omap_spinlock);
+		} else {
+			ret = -ENOMEM;
+		}
+	}
+
+	*olk_ret = olk;
+	return ret;
+}
+
+static bool omap_req_in_flight(struct scoutfs_lock *lock, struct scoutfs_omap_lock *olk)
+{
+	bool in_flight;
+
+	spin_lock(&lock->omap_spinlock);
+	in_flight = olk->req_in_flight;
+	spin_unlock(&lock->omap_spinlock);
+
+	return in_flight;
+}
+
+/*
+ * Make sure the omap_lock on the cluster lock is current.  We hold the
+ * lock while the request is in flight so the lock's write_version can't
+ * change while we're waiting.
```

Comment:
The cluster lock, where the write_version is.  write_version is changed every time a new write mode is acquired implying that another write could have been held since the previous write_version and the data protected by the lock could have changed.   I updated the comment.